File: bootcamp/week-18/3-proyecto/starter/item_service.py
class ExternalNotifier:
    def send(self, item_id: str, status: str) -> None:
        return None


class ItemService:
    def __init__(self, notifier: ExternalNotifier):
        self.notifier = notifier

    def activate(self, item_id: str, is_blocked: bool) -> str:
        if is_blocked:
            raise PermissionError("blocked item")

        self.notifier.send(item_id, "active")
        return "active"

File: bootcamp/week-10/3-proyecto/starter/item.service.js
class ItemService {
  constructor(itemRepository, billingGateway, auditLogger) {
    this.itemRepository = itemRepository;
    this.billingGateway = billingGateway;
    this.auditLogger = auditLogger;
  }

  async create(input) {
    if (!input?.name) {
      throw new Error("name is required");
    }

    return this.itemRepository.create(input);
  }

  async activate(itemId, planPrice) {
    if (!itemId) {
      throw new Error("itemId is required");
    }

    const item = await this.itemRepository.findById(itemId);
    if (!item) {
      throw new Error("item not found");
    }

    const chargeResult = await this.billingGateway.charge(planPrice);
    if (!chargeResult.approved) {
      throw new Error("payment rejected");
    }

    await this.itemRepository.markAsActive(itemId);
    this.auditLogger.log("ITEM_ACTIVATED", itemId);

    return { status: "active", itemId };
  }
}

module.exports = { ItemService };
